#11580 Handle trial logfile with directory separator

exarkun opened the pull request:
## Scope and purpose

Fixes #11580
Fixes #11657

exarkun committed:
Add `fileContents` matcher for asserting file contents

exarkun committed:
Handle missing parent directories for temp-directory and logfile

exarkun committed:
Annotate successResultOf correctly

exarkun committed:
Handle log files with directory separators in disttrial

twm commented on src/twisted/trial/test/matchers.py:
```diff
@@ -21,17 +23,46 @@ class _MatchAfter(BaseMatcher[_A]):
+            super().describe_mismatch(item, mismatch_description)
+        else:
+            mismatch_description.append_text(
+                f"{self.f.__qualname__}({item!r}) raised\n    {self._e}"
```
This should include the exception name, not just the message.

exarkun replied:
Done.  I ended up taking the message _out_ of the assertion made by the test (and only using the exception class name) because the message could vary based on locale.

twm commented on src/twisted/trial/test/test_matchers.py:
```diff
@@ -0,0 +1,82 @@
+            AllOf(
+                contains_string("fileContents"),
+                contains_string("No such file or directory"),
+                contains_string(p.path),
```
It looks like you're getting the repr of this in the Windows failure.

exarkun replied:
Oops.  I guess it's the repr on all platforms but it only makes a difference on Windows because of the platform sep there.  Took me a while. :/

twm commented on src/twisted/trial/_dist/test/test_disttrial.py:
```diff
@@ -304,6 +315,40 @@ def test_join(self):
+                # was found to be in-use already so we don't add a "/" suffix.
+                starts_with(config.workingDirectory.path),
+                # This should be exactly the suffix so we add a "/" prefix.
+                ends_with("/" + logFile),
```
Looks like you need `os.sep` here for Windows.

twm commented on src/twisted/trial/test/test_script.py:
```diff
@@ -33,6 +41,99 @@ def sibpath(filename):
+    return config
+
+
+runFromArguments = compose(runFromConfig, parseArguments)
```
_Technically_ I suppose this is a function definition that needs a docstring.

exarkun replied:
is there a meaningful way to give it a docstring?  I've seen people write things like:

```
x = ...
"""
Blah blah blah
"""
```

I know that Python itself doesn't do anything constructive with this but maybe one of the doc tools Twisted is using does?

Alternatively, maybe `compose` should synthesize a docstring based on its arguments? :slightly_smiling_face:

exarkun replied:
I filed https://github.com/twisted/twisted/issues/11659

twm reviewed (approved):
This looks like a solid fix, thank you!

Hamcrest makes the PR more verbose but the logic is quite easy to review. I'm curious how it impacts runtime.

I believe that the remaining failures will be trivial to address, so feel free to land once green!

exarkun commented:
> Hamcrest makes the PR more verbose but the logic is quite easy to review. I'm curious how it impacts runtime.

Regarding runtime - I think it's probably negligible?  I'm sure it's slower than `assertEqual`, maybe even measurably slow, but if it were even an extra millisecond on recent hardware I'd be quite shocked.

Hypothesis is much more likely to have a noticable impact on runtime.  That's part of the reason I opted for `sampled_from` with a short list here.  This substantially constrains Hypothesis' example search and limits the number of test iterations to a small handful (eg 4 for `test_logFile`).  This gives up a bit of Hypothesis' power by not letting it search out exciting values from the full input space and I'm not _sure_ it's the right balance to strike but it sticks closer to Twisted's historical testing patterns than the alternative so it seems like a good choice for now (in these early days of Hypothesis-in-Twisted).

exarkun commented:
> Hamcrest makes the PR more verbose but the logic is quite easy to review. I'm curious how it impacts runtime.

Regarding runtime - I think it's probably negligible?  I'm sure it's slower than `assertEqual`, maybe even measurably slow, but if it were even an extra millisecond on recent hardware I'd be quite shocked.

Hypothesis is much more likely to have a noticable impact on runtime.  That's part of the reason I opted for `sampled_from` with a short list here.  This substantially constrains Hypothesis' example search and limits the number of test iterations to a small handful (eg 4 for `test_logFile`).  This gives up a bit of Hypothesis' power by not letting it search out exciting values from the full input space and I'm not _sure_ it's the right balance to strike but it sticks closer to Twisted's historical testing patterns than the alternative so it seems like a good choice for now (in these early days of Hypothesis-in-Twisted).

exarkun committed:
Match the exception class name instead of its string representation

It would be nice to match against the string representation as well but
localization means we don't actually reliably know what the string will be.

exarkun committed:
Replace literal `"/"` with `os.dirsep` for cross-platform compat

exarkun committed:
Match the repr of the path in the exception report

If it has unprintables or special characters we'd rather have an unambiguous
interpretation, I guess.

Also, the builtin exception type does the string formatting so it's not like I
have a choice.

exarkun committed:
Merge remote-tracking branch 'origin/trunk' into 11580-trial-logfile-dirsep
